chord_metadata_service/restapi/models.py

```diff
@@ -1,4 +1,80 @@
+import logging
+from abc import abstractmethod
+from django.apps import apps
+from django.db import models
+from django.db.models import Q
+from django.core.exceptions import ValidationError
+from jsonschema import Draft7Validator
+
+logger = logging.getLogger(__name__)
+
+
 class IndexableMixin:
     @property
     def index_id(self):
         return f"{self.__class__.__name__}|{self.__str__()}"
+
+
+class SchemaType(models.TextChoices):
+    PHENOPACKET = "PHENOPACKET", "Phenopacket"
+    BIOSAMPLE = "BIOSAMPLE", "Biosample"
+    INDIVIDUAL = "INDIVIDUAL", "Individual"
+
+
+class BaseExtraProperties(models.Model):
+    extra_properties = models.JSONField()
+
+    class Meta:
+        abstract = True
+
+    @property
+    @abstractmethod
+    def schema_type(self) -> SchemaType:
+        """
+        Returns the SchemaType of the model.
+        Template method design pattern, implementation left to inheritors.
+        """
+        pass
+
+    @abstractmethod
+    def get_project_id(self) -> str:
+        """
+        Returns the Project.identifier of the project that owns this object.
+        Template method design pattern, implementation left to inheritors.
+        """
+        pass
+
+    def get_json_schema(self):
+        """
+        Returns a QuerySet[ProjectJsonSchema] containing schemas to validate
+        Template method design pattern, uses concrete defs of schema_type 
+        and get_project_id.
+        """
+        _project_id = self.get_project_id()
```

Comment:
underscore up front normally indicates an unused or private field, but this is used & can just be named like a normal local variable